application-collabora-default/src/main/java/com/xwiki/collabora/internal/rest/DefaultWopi.java

```diff
@@ -128,40 +125,44 @@ public Response getContents(String fileId, String token) throws XWikiRestExcepti
     public Response postContents(String fileId, String token, byte[] body) throws XWikiRestException
     {
         if (fileTokenManager.isInvalid(token)) {
-            return Response.status(Response.Status.UNAUTHORIZED).build();
+            throw new WebApplicationException(Response.Status.UNAUTHORIZED);
         }
 
-        AttachmentReference attachmentReference = this.attachmentReferenceResolver.resolve(fileId);
-        XWikiAttachment attachment = createOrUpdateAttachment(attachmentReference, body);
+        XWikiAttachment attachment =
+            attachmentManager.createOrUpdateAttachment(this.attachmentReferenceResolver.resolve(fileId), body);
 
         JSONObject response = new JSONObject();
         response.put(LAST_MODIFIED_TIME, dateFormat.format(attachment.getDate()));
 
         return Response.status(Response.Status.OK).entity(response.toString()).type(MediaType.APPLICATION_JSON).build();
     }
 
-    private XWikiAttachment createOrUpdateAttachment(AttachmentReference attachmentReference, byte[] content)
-        throws XWikiRestException
+    @Override
+    public Token getToken(String fileId) throws XWikiRestException
     {
         XWikiContext xcontext = this.contextProvider.get();
-        XWiki xwiki = xcontext.getWiki();
-        DocumentReference documentReference = attachmentReference.getDocumentReference();
-
         try {
-            // We clone the document because we're going to modify it and we shouldn't modify the cached instance.
-            XWikiDocument document = xwiki.getDocument(documentReference, xcontext).clone();
-            XWikiAttachment attachment =
-                document.setAttachment(attachmentReference.getName(), new ByteArrayInputStream(content), xcontext);
-            attachment.setAuthorReference(xcontext.getUserReference());
-
-            document.setAuthorReference(xcontext.getUserReference());
-            xwiki.saveDocument(document,
-                this.contextualLocalizationManager.getTranslationPlain("collabora.save.comment"), xcontext);
-
-            return attachment;
-        } catch (XWikiException | IOException e) {
-            throw new XWikiRestException(
-                String.format("Failed to create or update the attachment [%s].", attachmentReference), e);
+            String urlSrc = discoveryManager.getURLSrc(fileId);
+
+            Token token = (new ObjectFactory()).createToken();
+            token.setUrlSrc(urlSrc);
+            token.setValue(fileTokenManager.getToken(xcontext.getUserReference().toString(), fileId).toString());
```

Comment:
This ``getUserReference().toString()`` is not very clean. Couldn't you pass the user reference and let the file token manager serialize it using a proper serializer?